fix(transformer): exponentiation transform: do not replace object when private property

overlookmotel opened the pull request:
Fix exponentiation operator transform to bail out early if a private class property.

We can't transform this:

```js
class C {
    #p;
    method() {
        this.#p **= 2;
    }
}
```

But we should at least leave it alone. Previously `get_obj_ref` called `ast.move_expression(expr)` on the member expression's object before bailing out, so example above was transformed to:

```js
class C {
    #p;
    method() {
        null.#p **= 2; // <-- `null`
    }
}
```

This PR makes it spot this case early and bail out *before* calling `ast.move_expression(expr)`.

overlookmotel commented:
* **#6313** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6313?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
I thought this logic is part of that super weird secondary visit we discovered in babel with @Dunqing

overlookmotel commented:
> I thought this logic is part of that super weird secondary "environment visitor" we discovered in babel with @Dunqing

I wasn't completely sure what that "environment visitor" thing was, but my impression is that it's just a perf optimization.

What's meant to happen is it throws an error if you've not also enabled the class private properties transform. But we can't do that at the moment, as we've not implemented class transforms yet.

But what we definitely *shouldn't* be doing is raising no error and silently converting `obj.#p` into `null.#p`! That's what this PR deals with.

overlookmotel committed:
fix(transformer): exponentiation transform: do not replace object when private property (#6313)

Fix exponentiation operator transform to bail out early if a private class property.

We can't transform this:

```js
class C {
    #p;
    method() {
        this.#p **= 2;
    }
}
```

But we should at least leave it alone. Previously `get_obj_ref` called `ast.move_expression(expr)` on the member expression's object before bailing out, so example above was transformed to:

```js
class C {
    #p;
    method() {
        null.#p **= 2; // <-- `null`
    }
}
```

This PR makes it spot this case early and bail out *before* calling `ast.move_expression(expr)`.
